crates/bevy_asset/src/asset_server.rs

```diff
@@ -132,11 +132,27 @@ impl AssetServer {
         &self,
         path: P,
     ) -> Result<Arc<Box<dyn AssetLoader>>, AssetServerError> {
-        path.as_ref()
-            .extension()
-            .and_then(|e| e.to_str())
+        // Get filename
+        let s = path
+            .as_ref()
+            .file_name()
             .ok_or(AssetServerError::MissingAssetLoader(None))
-            .and_then(|extension| self.get_asset_loader(extension))
+            .and_then(|x| x.to_str().ok_or(AssetServerError::MissingAssetLoader(None)))?;
+
+        // If the filename has multiple dots, try to match on a multi-dot file extension
+        if s.split('.').count() > 1 {
+            if let Ok(loader) = self.get_asset_loader(&s[s.find('.').unwrap() + 1..]) {
+                return Ok(loader);
+            }
+        }
+
+        // Just get the file extension normaly if that doesn't work
+        let extension = &s[s
+            .rfind('.')
+            .ok_or(AssetServerError::MissingAssetLoader(None))?
+            + 1..];
+
+        self.get_asset_loader(extension)
```

Comment:
```suggestion
        let mut ext = s;
        while let Some(idx) = ext.find('.') {
            ext = &ext[idx + 1..];
            if let Ok(loader) = self.get_asset_loader(ext) {
                return Ok(loader);
            }
        }

        Err(AssetServerError::MissingAssetLoader(None))
```

For `foo.bar.sprite.png` this would try `bar.sprite.png`, `sprite.png` and then `png` as extensions to get a loader for.

Reply:
Didn't see https://github.com/bevyengine/bevy/pull/1277#issuecomment-764881281 before, but this seems cleaner to me.

Reply:
I agree, but it seems like we'd want to to preserve the e.g. `AssetServerError::MissingAssetLoader("png")` from get_asset_loader

Reply:
I added a test over in my branch for that, if that is indeed the behavior we want.

https://github.com/rparrett/bevy/commit/bd8639c6b64874fcd53f73e4c0e460307cc0c8bb

Reply:
Here's a modification of @rparrett's version that is nicely readable unlike my initial version and leaves great error messages. :smiley:

```rust

let s = path
            .as_ref()
            .file_name()
            .ok_or(AssetServerError::MissingAssetLoader(None))?
            .to_str()
            .ok_or(AssetServerError::MissingAssetLoader(None))?;

        // given the filename asset-v1.2.3.png, produce an iterator of file extensions that looks
        // like this: 2.3.png, 3.png, png
        let extensions = s.match_indices('.').map(|(i, _)| &s[i + 1..]);

        // Map extensions to loaders for each extension
        let loaders = extensions.clone().map(|ext| self.get_asset_loader(ext));

        // Default the result to an error if a mathching extension cannot be found
        let mut result = Err(AssetServerError::MissingAssetLoader(Some(format!(
            "Could not find asset loader for any of these file extensions: {}",
            extensions.collect::<Vec<_>>().join(", ")
        ))));

        // search for any successful asset loader, stopping at the first successful
        for loader_result in loaders {
            if loader_result.is_ok() {
                result = loader_result;
            }
        }

        // Return the final result
        result
```

Combine that with the tests and I think we're in good shape.

Reply:
I think that the inner value of MissingAssetLoader is meant to be the extension rather than an error message. `#[derive(Error)]` handles formatting the message, right?

It's just unfortunate that right now that the extension is not displayed there. I'm not sure if you can actually get the inner value of the optional string from the `#[error()]` attribute though.

Reply:
I do agree that the direction bjorn is going with the extension checking seems way cleaner though.

Reply:
> It's just unfortunate that right now that the extension is not displayed there. I'm not sure if you can actually get the inner value of the optional string from the #[error()] attribute though.

This should do it, I think. It's ugly, though.

```rust
#[error("no AssetLoader found{}", .0.as_ref().map(|x| format!(" for the following extension(s): {}", x)).unwrap_or(String::new()))]
    MissingAssetLoader(Option<String>),
```

> I do agree that the direction bjorn is going with the extension checking seems way cleaner though.

Yep, you're right. That makes more sense. I had missed that. I need to pay more attention. :smiley:

Reply:
Ah, this is better:

```rust
/// Errors that occur while loading assets with an AssetServer
#[derive(Error, Debug)]
pub enum AssetServerError {
    #[error("asset folder path is not a directory")]
    AssetFolderNotADirectory(String),
    #[error("no AssetLoader found{}", format_missing_asset_ext(.0))]
    MissingAssetLoader(Option<String>),
    #[error("the given type does not match the type of the loaded asset")]
    IncorrectHandleType,
    #[error("encountered an error while loading an asset")]
    AssetLoaderError(anyhow::Error),
    #[error("`PathLoader` encountered an error")]
    PathLoaderError(#[from] AssetIoError),
}

fn format_missing_asset_ext(ext: &Option<String>) -> String {
    return ext
        .as_ref()
        .map(|x| format!(" for the following extension(s): {}", x))
        .unwrap_or(String::new());
}
```